File: coinbase/iterators.py

from collections import deque

class BasicIterator:
    def __init__(self, arr):
        self.list = arr
        self.index = 0
    
    def hasNext(self):
        return self.index < len(self.arr) - 1
    
    def next(self):
        if not self.hasNext():
            raise RuntimeError("Accessing index out of bounds")
        self.index += 1
        return self.arr[self.index]

class RangeIterator:
    def __init__(self, start, end, step):
        if step == 0:
            raise ValueError("step cannot be zero")
        self.start = start
        self.end = end
        self.step = step
        self.current_value = start - step
    
    def hasNext(self):
        return self.current_value + self.step <= self.end if self.step > 0 else self.current_value + self.step >= self.end
    
    def next(self):
        if not self.hasNext():
            raise RuntimeError("Accessing value out of range")
        self.current_value += self.step
        return self.current_value

class InterLeavingIterator:
    def __init__(self, lists):
        self.queue = deque()
        for lst in lists:
            if lst:
                self.queue.append(iter(lst))

    def has_next(self):
        return bool(self.queue)

    def next(self):
        if not self.has_next():
            raise StopIteration()
        it = self.queue.popleft()
        val = next(it)
        try:
            # Try peeking the next item
            peek = next(it)
            self.queue.append(self._prepend(peek, it))
        except StopIteration:
            pass
        return val

    def _prepend(self, val, iterator):
        # Prepend a value to an iterator
        yield val
        yield from iterator
    
lists = [[0, 1, 2], [], [3, 4], [5]]
zigzag = InterLeavingIterator(lists)

result = []
while zigzag.has_next():
    result.append(zigzag.next())

print(result)  # Output: [0, 3, 5, 1, 4, 2]

File: coinbase/iterators.cpp
#include<iostream>
#include<vector>
#include<queue>

using namespace std;

class Iterator {
public:
    virtual bool hasNext() = 0;
    virtual int next() = 0;
    virtual ~Iterator() = default;
};

class BasicIterator : public Iterator {
private:
    vector<int> data;
    int currIdx;

public:
    BasicIterator(vector<int>& data) : data(data), currIdx(0) {}

    bool hasNext() override {
        return currIdx < data.size();
    }

    int next() override {
        if(!hasNext()) {
            throw runtime_error("Out of bounds access");
        }
        return data[currIdx++];
    }

};

class RangeIterator : public Iterator { // [start, ... end)
private:
    int start, end, step, currVal;

public:
    RangeIterator(int start, int end, int step) {
        if(step == 0) {
            throw runtime_error("Step size cannot be 0");
        }
        this->start = start;
        this->end = end;
        this->step = step;
        this->currVal = start;
        if((step > 0 && start > end) || (step < 0 && start < end)) {
            this->currVal = end;
        }
    }

    bool hasNext() override {
        return step > 0 ? currVal < end : currVal > end;
    }

    int next() override {
        if(!hasNext()) {
            throw runtime_error("out of range access");
        }
        int ans = currVal;
        currVal += step;
        return ans;
    }
};

class ZigZagIterator {
private:
    queue<pair<vector<int>::iterator, vector<int>::iterator>> queuedData;

public:
    ZigZagIterator(vector<vector<int>>& data) {
        for(auto& vec : data) {
            if(vec.size() != 0) {
                queuedData.push({vec.begin(), vec.end()});
            }
        }
    }

    bool hasNext() {
        return !queuedData.empty();
    }
    
    int next() {
        if(!hasNext()) {
            throw runtime_error("no next element present");
        }
        auto [it1, itEnd] = queuedData.front();
        queuedData.pop();
        int val = *it1;
        it1++;
        if(it1 != itEnd) {
            queuedData.push({it1, itEnd});
        }
        return val;
    }
};

class ZigZagIteratorOfIterators {
    queue<Iterator*> queuedIters;

    ZigZagIteratorOfIterators(vector<Iterator*>& iterators) {
        for(auto iter : iterators) {
            if(iter->hasNext()) {
                queuedIters.push(iter);
            }
        }
    }

    bool hasNext() {
        return !queuedIters.empty();
    }

    int next() {
        auto it = queuedIters.front();
        queuedIters.pop();
        int val = it->next();
        if (it->hasNext()) {
            queuedIters.push(it);
        }
        return val;
    }
};

int main() {
    vector<int> data1 = {1, 2, 3, 4, 5};
    BasicIterator basicIter = BasicIterator(data1);
    while(basicIter.hasNext()) {
        cout << basicIter.next() << " ";
    }
    cout << endl;

    RangeIterator rangeIter1 = RangeIterator(1, 10, 1);
    while(rangeIter1.hasNext()) {
        cout << rangeIter1.next() << " ";
    }
    cout << endl;

    RangeIterator rangeIter2 = RangeIterator(1, 10, -1);
    while(rangeIter2.hasNext()) {
        cout << rangeIter2.next() << " ";
    }
    cout << endl;

    RangeIterator rangeIter3 = RangeIterator(1, 10, 2);
    while(rangeIter3.hasNext()) {
        cout << rangeIter3.next() << " ";
    }
    cout << endl;

    RangeIterator rangeIter4 = RangeIterator(10, 1, 1);
    while(rangeIter4.hasNext()) {
        cout << rangeIter4.next() << " ";
    }
    cout << endl;

    RangeIterator rangeIter5 = RangeIterator(10, 1, -2);
    while(rangeIter5.hasNext()) {
        cout << rangeIter5.next() << " ";
    }
    cout << endl;

    vector<vector<int>> data2 = {
        {1, 2, 3, 4, 5},
        {6, 7, 8},
        {},
        {9, 10, 11, 12}
    };
    ZigZagIterator zigZagIter = ZigZagIterator(data2);
    while(zigZagIter.hasNext()) {
        cout << zigZagIter.next() << " ";
    }
    cout << endl;
}